### ISSUE TITLE
Discussion: Language release cadence

### ISSUE THREAD

**Issue by HaloFour** [dotnet/csharplang#2518]
I've been thinking a little about the C# release cadence lately and how it relates to how and when features get shipped.  I recently listened to a talk from Brian Goetz where he discussed Java's recent adoption of a consistent release cadence, why they decided to do this and the benefits they found from doing so and I think it's worth having the discussion as to whether or not the same could apply to C#.

Previously Java would tie major features to major releases.  Those releases would be relatively infrequent, on the order of one every few years, and because those major language features were the driving force behind the release it created a complex project management cycle.  It also led to, and I'm going to paraphrase this very poorly, a "constipation of smaller features", as the big release-driving features would get all of the attention.  The releases also tended to be big which increased risk and complexity for any user choosing to update.

As of Java 9, they now follow a strict six-month release cycle.  There's no such thing as a "major" release anymore, each release is an increment of the major version number.  Java 9 released in Sept 2017, Java 10 released in March 2018, Java 11 released in Sept 2018, Java 12 released in March 2019 and Java 13 will release in Sept 2019.

As described, any language features ship with the compiler when they're ready.  For smaller features that are not controversial that means as soon as they're done they are ready and included in the next release.  For larger features that may be developed over the span of multiple releases, the feature will ship behind an "experimental" flag.  If that feature is well received as is it will be ungated in the next release, or it could be tweaked and left behind the flag for another release, or it could be pulled altogether.

The benefits here should be pretty obvious to any team who has worked in a continuous delivery environment.  Features come quick and often.  There isn't a constant focus on the big-bang features that define a release.  And there is a lot less of a focus on the release in general.  Project management is still a thing, of course, but it's less tied to and obsessed about proposed release dates.  Brian Goetz specifically mentioned that after adopting this cadence the time he had to spend in release planning meetings dropped to almost nil and he could spend most of his energy on the language, where it belongs.  I don't know what @MadsTorgersen does all day right now, but it wouldn't surprise me if meetings were towards the top of the list.

So, what I propose, is that the C# team drop the notion of major and minor releases.  Exactly how the numbering would work can be up for discussion, but drop the idea that big features are a part of major releases and small features or tweaks are a part of minor releases.  Each release should stand on its own, and each release should come on a fairly consistent schedule.  If that's at all possible I would also like to see the adoption of a release process like that mentioned for Java: a feature releases when it's ready into the next release, not when enough features can be bundled together with a big feature to justify incrementing the number.  And, importantly, drop the notion of "preview" releases leading up to a major release.  Allow "preview" versions of upcoming features to flow into the normal release cadence.

I do acknowledge that there are numerous differences between C# and Java which will complicate these ideas.  C# has dependencies on separate teams building the CLR whereas Java always evolves in lockstep with its runtime.  C# also has different concerns around the IDE and tooling.  I don't remotely pretend to have answers around how these issues can be resolved.

Link to Brian Goetz discussing the changes to Java release cadence, starting at the relevant bit:
https://youtu.be/4r2Wg-TY7gU?t=475

**Comment by HaloFour** [comment]
Oh crap, that turned into a diatribe.  Anyway, this is intended more as an open discussion.  I'm not proposing anything specific.  However, listening to Brian Goetz discuss this in detail recently left an impression on me.  If the lumbering dinosaur that is Java can adopt a CD-like release cadence, and see real benefits from doing so then perhaps we can learn something from this.

**Comment by svick** [comment]
Personally, I think numbering language versions based on features makes sense, because it makes it much easier to figure out what's going on. There are already way too many versions in .Net (.Net Core SDK 2.2.106, .Net Core Runtime 2.2.4, C# 7.3, Roslyn 2.10, MSBuild 15.9.20, VS 15.9.11), I'm not sure forcing incrementing the version of C#, seemingly independently of other parts of the ecosystem, would be a good idea.

>  Features come quick and often.

How come Java's 2 releases per year are "quick and often", but C#'s 3 releases per year in 2017 are not? (Though 2018 had only one release and 2019 is probably not going to have more than 2 at this point.)

> Allow "preview" versions of upcoming features to flow into the normal release cadence.

How is that different from what's currently happening, where VS 2019/Roslyn 3.0 contain a preview version of C# 8.0?

**Comment by HaloFour** [comment]
@svick 

> Personally, I think numbering language versions based on features makes sense, because it makes it much easier to figure out what's going on.

The features can still be tied to a version.  They're just not tied to **that** specific version, nor is there an expectation that because a major version number may be incrementing that it has to be accompanied major.

> How come Java's 2 releases per year are "quick and often", but C#'s 3 releases per year in 2017 are not?

Java's are consistent.  C# had a burst and it's clogged up again, pending a major release.  I expect that after C# 8.0 releases we might get a quick burst of minor releases, but the cycle would likely continue that once work really starts on the next major features that nothing new will be released.

> How is that different from what's currently happening, where VS 2019/Roslyn 3.0 contain a preview version of C# 8.0?

It's not C# 8.0, nor is it C# 7.3.  It's a strange in-between with it's own release cadence and expectations.  And while I would fully expect that the team would be willing to delay any of those features if something wrong shook out in the preview, the expectation is that those major features are going to be tied to that major release.  The result of the preview isn't going to be C# 7.next, it's going to be C# 8.0, and that sets everyone's expectations, the users, the developers, the project managers, etc.

**Comment by jaredpar** [comment]
Looking at the [.NET 5](https://devblogs.microsoft.com/dotnet/introducing-net-5/) blog post you'll see that .NET is moving to regular release cycles. That's not just the runtime, it's the suite of tools in the .NET SDK that are moving to this cycle.

That includes the compiler. While I can't promise that every release of .NET will have a corresponding C# release it seems like that it will be the norm to do so and such releases will be major ones. 

> So, what I propose, is that the C# team drop the notion of major and minor releases. 

I expect that minor release will be heavily de-emphasized in this model but they're still a valuable tool for us to have around when needed. The way I view minor releases is that we can continue to use them when circumstances call for it. Such as the need to support a critical scenario that comes up in between major releases of .NET. 

That seems increasingly unlikely with the rapid release plan of .NET. But if it did happen I'd feel fine using a minor release to deliver the feature.

**Comment by HaloFour** [comment]
Here's a thought exercise:  Let's say that during this preview that it's found that there are too many issues with NRTs and DIM for it to ship this year.  Would the next release of C# still be 8.0?  What would the thoughts be on shipping those otherwise major features in minor releases?  Let's assume whatever holds up DIM isn't in .NET Core 3.0 as that would complicate the scenario considerably.

**Comment by jaredpar** [comment]
> Let's say that during this preview that it's found that there are too many issues with NRTs and DIM for it to ship this year. Would the next release of C# still be 8.0? 

Absolutely. This is a new version of C# that is shipping in tandem with a release of the .NET SDK. Those releases will be major releases, even if a few big features ultimately missed the cut. 

Features missing the cut is rare but not unheard of. It's happened before, it will happen again. In the past we've adopted the stance I listed above.

**Comment by HaloFour** [comment]
@jaredpar 

That's pretty cool to hear.  Perhaps everything I've described would shake out of the annual release cadence.  I, personally, would like to see language releases a little more frequently, like every six months, as that would allow a for faster adoption of smaller features.  Perhaps that's where minor releases fit, although I'd prefer that would emphasize the timing and not the size or quality of the features included.  Ideally those SDK releases also wouldn't be held up even if features start to slip.

**Comment by MgSam** [comment]
Forcing release cycles does get stuff out there more quickly- but it can also have converse effect of what you describe- because your dev cycles are relatively short you never ship major features because there's never enough time to develop them properly. 

IMHO, this is something that happened to Windows when they decided to do a release every 6 months. Windows 10 RTM in 2015, and since then every update has been strictly minor changes. To be honest, I can't name a single feature (other than endless bugs) that has been added since Windows 10's release.

I think C#'s recent release cycle has almost hit the sweet spot. Minor releases ship small features, major releases ship major features which may require runtime support. If anything, the minor releases have even happened a little too frequently (the `in` keyword being the perfect example of a rushed feature that is almost universally panned and now stuck in the language forever). 

I'd prefer they stick to a "when it's done" model of major/minor versions and not force yearly releases to meet artificial deadlines.

**Comment by HaloFour** [comment]
@MgSam

> because your dev cycles are relatively short you never ship major features because there's never enough time to develop them properly.

The answer there is to not tie the development cycle to the release cycle.  Projects may span many releases.  Whenever they happen to be done they get slated to go out with the next release.  The release itself can never be held up, neither can any of the other myriad of changes already ready go out.

There have been numerous Java features which have been released in this manner.  The new HttpClient took 2-3 releases.  There are two GCs currently in experimental mode, as well several language features.  There are a number of long term projects still under active development which aren't likely to be included in any releases this year, but may trickle out piecemeal as experimental.  Some of those projects have been under active development for over 5 years.  The faster release cadence doesn't seem to be discouraging the major projects, they just deemphasize their timing and don't block/slow the constant cycle of other new features.

Sure, it could be done wrong.  It could also be done right.  We're watching the C# language deal with fits and starts when it comes to releasing language features.  There's lot of small features that are currently stuck waiting on C# 8.0.  And with everyone head down on those major features to meet the arbitrary deadline we all experience the relative silence from the team.  This happened before C# 6.0 and C# 7.0 as well.  With a faster release cadence we'd theoretically see those smaller features sooner and we'd have continuous feedback regarding the development of features into the next cycle.

**Comment by jaredpar** [comment]
> The answer there is to not tie the development cycle to the release cycle. Projects may span many releases

Bingo.

This already happens today. Consider that we started implementing nullable reference types in 2015 for C# 7.0. Spent ~3 months on the initial parsing and symbol model. After that though the C# 7.0 release began to take shape, as did our costing for NRT and we decided to delay it because it obviously did not fit. Once C# 7.2 was wrapping up we shifted resources back into NRT and are pushing it to completion for C# 8.0.

Another example to site here is pattern matching. The feature set we're releasing for C# 8.0 is actually fairly close to the original vision that we had back in the C# 7.0 time frame. Once we got into the implementation though we realized there wasn't enough time to fit it in. Hence we split it into two pieces, each of which is a large feature onto itself. If there hadn't been a logical split then most likely we would've just kept working on it and targeted it all for C# 8.0.

**Comment by MgSam** [comment]
> The answer there is to not tie the development cycle to the release cycle.

I think that's easier said than done depending on the pressures on the team. Often when you're forced to plan releases to meet some arbitrary deadline, it means you are constantly doing exactly work: re-prioritizing so that _something_ is in your releases as well as all the inevitable extra work that comes with having a release.  Which means the big things tend to get pushed aside so you can complete a bunch of little things on schedule.

I agree there are lot of minor features in C# 8 that could have been shipped sooner. There is an argument that maybe there should have been a C# 7.4 in the interim. But I assume the reason they didn't do that is because it would have required lots of release planning and whatnot that would have distracted from the big features they wanted to do for C# 8- exactly my point. We may have gotten `??=` sooner but NRTs may have been delayed another year.

**Comment by jmarolf** [comment]
I think the goals expressed here by @HaloFour  are, as @jaredpar says, the C# teams goals.
- do not tie development cycle and release cycle together
- try and not do a big bang release every 3 years at the expense of other smaller-but-still-useful features

And we are trying to do these things.  The reality is that the gravitational pull of large features will always have some effect on smaller features. In perfect world we would just have Chuck working on nullable references constantly and that would not effect anyone else, but in reality some things just need variable amounts of brain power over their development cycle to complete. 

We are always trying to find better ways to work/schedule things and I love this conversations.  I just don't know if its possible to live in a world were large features have no effect on the development of smaller ones...

**Comment by HaloFour** [comment]
@MgSam 

> Often when you're forced to plan releases to meet some arbitrary deadline,

The point is to not plan the features around the arbitrary deadline.  The releases will happen either way.  What features end up in that release depend only and entirely on what features are "done".

> But I assume the reason they didn't do that is because it would have required lots of release planning and whatnot that would have distracted from the big features they wanted to do for C# 8- exactly my point.

And if the releases happen either way the release planning becomes infinitely easier.  You don't have to gauge and schedule the release based the perceived amount of work remaining to be done.  That kind of estimation is near impossible in this kind of work.  Instead the release planning is simply, "what's done?"  Brian Goetz very explicitly stated that the amount of time he had to spend in release planning meetings dropped considerably (paraphrasing, almost to none) compared to when they did big-bang releases.

Either way, it is certainly a question of balance.  My experience with C# is that things tend to cram for the big releases.  Even with minor versions in the mix the majority of the work seems to be focused on the larger features and towards these arbitrary deadlines we end up with relative silence on these forums.  That's been my impression since CodePlex, and it's certainly the case now leading up to C# 8.0.  And from the stories outlined it seems that when a big feature slips it doesn't do so to the next release, it does so to the next major version, so we're already seeing slippage as a result of this arbitrary alignment.

@jaredpar 

What I read into the two anecdotes you provide regarding NRTs and pattern matching is that when two big features didn't make it into a major release that their work become deprioritized to realign them with the next big release.  Perhaps I'm interpreting that wrong, but that's precisely what I think should be avoided.  Ideally, the work on both projects would continue on and if either happened to be done by C# 7.1 or C# 7.2 or whatever then they would be released at that time.  Granted, I understand that there are a lot of reasons to stop work on a particular project, including reprioritization.

@jmarolf 

I don't know that I have any "goals" per se.  I am sharing what I learned listening to Brian Goetz discuss  the changes to Java's development cycle, and what I perceive to be the dividends of that change as a Java programmer.

I found a video of Brian Goetz discussing this topic at a different conference.  I will post a link to the relevant portion above.

**Comment by marksmeltzer** [comment]
While this is definitely a meta discussion worth having, I can say that from the trenches, the current approach is working quite well.

I would be interested to see a more formal C# release cadence announcement sometime around the NET 5 release. But I am also very happy to have the C# releases to have its own timeline.

Having a release every X weeks/months is a very good Agile practice worth adopting. Only ship what's actually done. Have an active project dashboard that lists what's done and what's in progress. That keeps the community in the loop on what to expect in the next release ahead of time.

Have a build that includes only the done stuff, abd also have a preview build that includes work in progress of mixed quality that insiders can experiment with and provide feedback on. 

That's the best of all scenarios.

**Comment by vbcodec** [comment]
Best way - imo - is to release minor version when few small features are ready, and major version when one or more big features are ready, not excluding smaller features if they are also ready. All C# releases should be desynced from VS and NET releases, or any timely regular releases. 
Drawback is that some NET release may include improved runtime and libraries, and provided compiler won't use it, but it won't be disaster to wait few months when new C# version catch up these improvements.

**Comment by jmarolf** [comment]
> Ideally, the work on both projects would continue on and if either happened to be done by C# 7.1 or C# 7.2 or whatever then they would be released at that time. 

I guess that is what I am getting at. We can't do all the things all the time.  Some features will need to have a lot of people on them to get done and we need to budget and schedule for that. I think what we don't want is that pushing too many other small features out.  The other issue we need to think about it version numbers.  Like it or not people _really_ care about version numbers.  I don't know if people in the wider C# community would like us to release nullable reference type in C# 7.4 vs. 8.0.  Its all just a messaging thing, but we'd need to do a good job of communicating version number strategy.

**Comment by HaloFour** [comment]
@jmarolf 

> I guess that is what I am getting at. We can't do all the things all the time. Some features will need to have a lot of people on them to get done and we need to budget and schedule for that.

Sure, big projects may not get all of the attention all of the time.  But they should get the attention when there is time and energy to give it attention, not when the schedule aligns conveniently with a potential major release.

> The other issue we need to think about it version numbers. Like it or not people really care about version numbers.  I don't know if people in the wider C# community would like us to release nullable reference type in C# 7.4 vs. 8.0. Its all just a messaging thing, but we'd need to do a good job of communicating version number strategy.

Exactly, the distinction between major releases and other releases creates a perception and an expectation amongst everybody.  If everyone thinks that major releases must contain something grand then they get held up until something grand is ready, which forces release management to schedule and prioritize work around that idea.  Everyone gets too obsessed with when that major release will be released rather than releasing what's ready when it's ready.  And everything that is ready is stuck.

I think a lot of projects have demonstrated quite successfully that not only can you do away with major releases, but that it doesn't at all detract from when the major features get released.  Valhalla and Loom will both bring massive changes to the Java world, on the order of, if not bigger than, .NET 2.0 compared to .NET 1.0.  But it won't be met with a special number.  It'll be Java 15 or 16 or 17 or whatever happens to be the release at the time they have something to ship.  And because of the cadence they can focus on carving up the big projects into smaller chunks that can be released incrementally and/or experimentally.

**Comment by CyrusNajmabadi** [comment]
What woudl help me would be an explanation of how thins would be different, using the last couple of releases of C# as an example.  What would have been different in the numbering, what would have changed for users, and how would things have been different for the team if this changed approach had been taken.  

it's not clear to me how it would have affected all of hte above, and whether or not it would be positive across the board, or would have a range of positives and negatives that would have to be evaluated against each other.

**Comment by alrz** [comment]
Reiterating on what others said, I think the current planning is more about resource management rather than features tied to a specific version. Side that with all the other work in other areas, IDE, runtime, etc which all may need to meet the same deadline. Major versions define the focus of the team, so meanwhile there can be no release except if you want to split resources and push big features further back. 

I understand the idea here is to flat number all the releases and do parallel work on features, but considering that C# team is a small team I think the frequency of these releases may have a negative effect on *when* a feature will be released. 

That said, I think I'd prefer faster 7.1, 7.2, 7.3 and a late 8.0 release with the feature X, versus relatively slow but steady 8, 9, 10 releases which all include an experimental version of the feature X.

**Comment by DavidArno** [comment]
@MgSam,

> IMHO, this is something that happened to Windows when they decided to do a release every 6 months. Windows 10 RTM in 2015, and since then every update has been strictly minor changes. To be honest, I can't name a single feature (other than endless bugs) that has been added since Windows 10's release.

This is always the difficulty of categorizing major and minor features. The Windows team for example added the Linux kernel to Windows 10 recently. That is about as massive a change to windows as one could ever get ... if you use it of course. If you don't, it's somewhere between minor and irrelevant. Likewise, for me, DIMs is a minor feature (I'll likely make use of the ability to embed an internal class inside a public interface, but that's it). For others it's a real biggie. 

When you've waited two years for a change, even little features can be a big deal. When change happens all the time, it's easy to get blasé about things and to only ever see minor changes, if the changes are even noticed.

So I'd contend that frequent releases doesn't lead to "strictly minor changes", it just shifts our perception of what constitutes a major change.

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> What woudl help me would be an explanation of how thins would be different, using the last couple of releases of C# as an example. What would have been different in the numbering, what would have changed for users, and how would things have been different for the team if this changed approach had been taken.

For starters, we wouldn't go almost a year and a half between releases.  C# 7.3 was released in May 2018 and C# 8.0 is slated to be released in Sept 2019.  And all of the minor releases were focused on smaller tweaks to the language, so you can kind of look at it as two and a half years between releases.

As for the numbering, let's say C# ditched minor versions and every release incremented the major version.  I think that would deemphasize prioritization and allocation around the longer timeframes so that projects can be worked on when they're ready and when there are resources to work on them.  The experience from the user is a more constant and steady stream of both major and minor features.  The experience from the team, per Brian Goetz, is a lot less time spent in meetings trying to plan out what will fit into a major release timeframe.

@alrz 

>  I think the current planning is more about resource management rather than features tied to a specific version.

How much of that is an artifact of the major/minor release cycle?  From the anecdotes above it sounded like every major feature that didn't make the cut for C# 7.0 was deprioritized.  That sounds like the resource allocation and planning was based around a much more long term release cycle rather than when there were resources available to work on a given project.  What happens if one of those major features has to be punted again?  Do we have to wait another 2+ years for the next major release?

> I understand the idea here is to flat number all the releases and do parallel work on features, but considering that C# team is a small team I think the frequency of these releases may have a negative effect on when a feature will be released.

I can't comment on the size of the C# team compared to the Java team.  I don't even think I know how big the C# team is these days.  But I assume that the Java team similarly has much more work than they have people to work on it so they face the same issues of prioritization and resource allocation.

I'd probably be just as suspicious as the others here if I didn't think that this was going well for Java.  The experience from the team as related by their architect has been very positive and that the effort of the team can be focused much more of the language rather than on project management around release planning (that wasn't mentioned in the YouTube video I posted, I will post another once the video of the conference I attended is published).  The users get a steady stream of new features of varying sizes.  The difference has been transformative.

As for team velocity, Java 9 took 3.5 years and completed 91 JEPs.  Every release since took six months and completed between 10 and 20 JEPs.  If you average that out to around 30 JEPs per year, the team seems to be completing around the same amount of work, if not a little more.  That's assuming that all JEPs are equal, which of course they're not.

**Comment by HaloFour** [comment]
@DavidArno 

> So I'd contend that frequent releases doesn't lead to "strictly minor changes", it just shifts our perception of what constitutes a major change.

And quite a few major changes are really the culmination and combination of a bunch of orthogonal minor changes.  Are they less major if they are released incrementally over a period of time?

This is already the case with pattern matching.  We had a bunch of changes that shipped with C# 7.0 that fit under that umbrella, and we'll have another bunch shipping with C# 8.0.  But there's still more changes like records and DUs that also fall under pattern matching which will likely come later.  Is it any less major because it's being split over at least 3 releases?  Would it be any less major if those same features were incrementally metered out over 6 releases in a rapid cadence?  Would we recognize it as "pattern matching", the big feature, or instead as just a bunch of tiny disparate features?

**Comment by marksmeltzer** [comment]
@CyrusNajmabadi @jmarolf 

Just a question out of curiosity, is there a strong reason to avoid using semantic versioning? E.g. we might see C# 8.97 before we see a breaking change that would bring around C# 9... as a fan of semantic versioning, I would appreciate a system like that, but perhaps there's some good rationale to avoid it.

**Comment by HaloFour** [comment]
@marksmeltzer 

I'm not sure semantic versioning would make sense given that the language design team goes to great lengths to avoid breaking changes.  Not that breaking changes don't happen, but only if the potential of breaking someone's project is exceedingly small.  I wouldn't expect the C# team would adopt a versioning system intended to telegraph the message to expect incompatibilities between major releases.  If you had to meet a minimum threshold of breakage we'd probably still be on C# 1.x.

**Comment by marksmeltzer** [comment]
@HaloFour 

I certainly recognize that that has been the case. But I for one would like to see some breaking changes around the .NET 5.0 timeframe. I would love to see things the community wants removed flagged as obsolete and then eventually removed from the language. As long as the underlying CLI is backwards compatible, that wouldn't affect old assemblies already in the wild -- it would just affect active builds.

For instance, I would love to see the team use that approach to solve the NRTs conundrum, and eventually end up in a place where nullability is always explicit like TypeScript and Kotlin. Creating a formal mechanism, such as semantic version, for making breaking changes known would provide path to actually accomplish that goal.

One method could be to increase the major version when things are marked as obsolete (but will still compile with a compiler flag), and obsolete things can only be removed in the subsequent major version. So if major versions are spaced out fairly widely, that would provide the community ample time to adapt their active code bases.

Moreover, automated conversion tools for the new obsolete features could be included in the major releases -- for example, an automated tool for converting all legacy reference variables/fields to be explicitly nullable would be a way to bridge the gap to explicit NRTs.

**Comment by jmarolf** [comment]
@marksmeltzer 

> I would love to see things the community wants removed flagged as obsolete and then eventually removed from the language.

I can say definitively this will never happen. C# has too many users. We'd just create a new language if we were going to create sweeping breaking changes.  

@HaloFour 
> How much of that is an artifact of the major/minor release cycle? From the anecdotes above it sounded like every major feature that didn't make the cut for C# 7.0 was deprioritized. 

Internally we want the process to be that there are well understood shipping vehicles and things just move into them when they are ready. Visual Studio and .NEt Core (which are stillfactors in planning C#) are also trying to get better and better at shipping this same way.  I think we are all in agreement that this is a good thing. I just don't know if we can avoid big features sucking all the oxygen out of the room when they ramp up.

**Comment by marksmeltzer** [comment]
@jmarolf

I agree that old code should still compile. I don't see any reason that goal would be compromised by also depreciating obsolete language features over time.

The legacy constructs for C# should still compile if the project is pegged to a specific legacy version of C#.

That seems like the best of both worlds to me.

**Comment by CyrusNajmabadi** [comment]
> The legacy constructs for C# should still compile if the project is pegged to a specific legacy version of C#.

That's a larger maintenance burden.  And, if the language has to keep supporting these constructs, why not just allow people to keep using them?  

> One method could be to increase the major version when things are marked as obsolete (but will still compile with a compiler flag), and obsolete things can only be removed in the subsequent major version. So if major versions are spaced out fairly widely, that would provide the community ample time to adapt their active code bases.

Or, it will keep organizations never moving forward.  See, for example, python2 and 3.  These types of approaches risk large schisms in teh ecosystem.

**Comment by marksmeltzer** [comment]
@CyrusNajmabadi 

Deprecating unwanted language features seems to be working for TypeScript. Not all communities are equal. In C#'s case, I think the few things we'd want to jettison can easily be addressed with automatic conversion tooling (e.g. explicit NRTs). 

Sure there would be some maintenance burden on the compiler to support compiling old language versions -- but how much effort would it actually take quantitatively? Shouldn't be much. (I'm saying that with the authority of being someone who provides deep backwards compatibility for the frameworks I ship, and I also deprecate things using the strategy described above. So I do know a bit about what would be involved...)

**Comment by CyrusNajmabadi** [comment]
I'd have to see which features were even on hte list for deprecation.  Are there ones in particular that we think would even be up for this sort of thing?

**Comment by YairHalberstadt** [comment]
Which features in particular are negatively affecting the ability of the language to move forward.

For example, shorthand delegate syntax is now a bit pointless, but it's in, didn't cause any harm, and it's not worth the pain of removing.

Then there's custom casts, which make the whole language more complex, and which some language designers definitely wish hadn't been added. But they are to deeply embedded in the ecosystem to be removed.

So what's left?

**Comment by HaloFour** [comment]
I bet the team would love to deprecate and remove method overloading.  😀

**Comment by marksmeltzer** [comment]
@CyrusNajmabadi 

I am pretty happy with C# overall, and the things on the table for C# 8.0. The main use case I can perceive deprecation helping with would be a transition to full explicit NRT support, where nullability was always explicit. TypeScript and Kotlin have both provided syntax concepts for a lot of the necessary mechanisms -- like late initialization, etc.

I know the C# community is currently wrestling with C# 8.0's various proposals about how to shoe horn NRTs into existing C# code. And I agree with you, generally, that the cost/benefit needs to be "just right" if there aren't going to be any breaking changes to C#.

However, I would be fine with C# 8.0 containing an intentionally "transitionary" NRT implementation that is "friendlier" to existing code bases. Then perhaps C# 9.0 could make the hard break that I'm describing here, and provide tooling to auto-port old code to the new explicit syntax.

However, this discussion about deprecation doesn't have to be a scenario that forces old code not to compile. Rather, one way of implementing semantic versioning is to have a set of quirks, and the new version would come with certain quirks enabled/disabled by default. That's how the .NET framework handles functional deprecation and it is extremely pragmatic. Projects can upgrade to the newer framework and then opt in or out of the various quirks relevant to ensuring their code continues to function properly.

Consumers are advised to upgrade their code to avoid the uses of quirks modes -- but that is a pragmatic decision left to the consumers. The same general approach can be used for language design as well to provide another option between the two extremes of never deprecating anything and hard deprecation. (Moreover, if certain legacy quirks fade into obscurity they could be removed.)

Another glaring option to improve C# would be improving the code that the compiler generates for async scenarios. For instance, the F# and C++ teams had the privilege of working on their async implementations later in time, and there are now well known ways of doing it better.

Although @stephentoub and company have been making head way on providing pathways to avoid allocation in asynx scenarios using `ValueTask<T>`, it would still be really nice to have an alternative async subsystem in C# built. I don't know what the precise implications of `doing asynx better` would mean for C#, but that is currently set of a dead options due to the current backwards compatibility policy.

Third, I've discussed in other threads that I would like to see C# compiler treat non-capturing delegates and lambdas expressions as immutable and single instance. If there are certain concerns the team has about the underlying data structures wrt immutability, those could be addressed via breaking changes as well (e.g. compilation controlled by quirks mode, etc.).

The community might have other ideas that are better than the ones I've thought up.

**Comment by HaloFour** [comment]
@marksmeltzer 

At that point you might as well make a clean break and start a new language.  Then you have no need for "quirks mode" or varying dialects.  You're going to end up with a schism either way, you might as well make the most of it.

In any case, my opinion is that the idea of deprecation of off-topic and orthogonal to this particular issue.  I'd like to avoid the notion of major/minor language versions in general, not overload them to mean even more than they do today.

**Comment by marksmeltzer** [comment]
@CyrusNajmabadi 

Shorter answer: I think we'd need a survey.

**Comment by marksmeltzer** [comment]
@HaloFour 

I am working on a language of my own, and I'm planning on going the quirks mode route. 

Each file in my language begins with a simple version instruction, and then will also allow a set of flags to be defined to opt in/out of quirks. The version instruction is optional, and if not specified the lexer/parser assumes the latest version.

As a language designer, that gives me the freedom to make mistakes and correct them. Do I plan on deprecating things very often once I start shipping the tooling? Hopefully not. But in my prerelease iterations I have already made heavy use of that capability while I've been experimenting with various syntax formalizations, etc. I don't always want to rewrite my existing working code when I introduce a breaking change, so I'm using this capability in my own codebase to enable rapid cycles for the language itself.

(For my language, when I get ready for a first release, I'll squash all of the deprecated features and update all of my code to establish a consistent baseline. But I know I will appreciate having the glue already in place to enable pivoting for later releases, precisely because I've already been using it with great success.)

I like TypeScript's boldness of just releasing breaking changes when they seem like the right thing to do, but I don't like the idea of actually breaking working code.

**Comment by CyrusNajmabadi** [comment]
@marksmeltzer Sounds fascinating!  Would def love to hear how it works out for you :)

**Comment by jaredpar** [comment]
> Which features in particular are negatively affecting the ability of the language to move forward.

This is the key question and the answer is really ... nothing. There are occasions where existing behaviors cause us a bit of pain when working on the design for new features but it's on the level of annoyance, not a blocker. 

The most recent example is using the `_` syntax for discards. Unfortunately the compiler allows for `_` to be a legal identifier. This means when processing `_` we can't assume it's a discard but instead have to determine if it's a declared identifier and failing that then move onto processing it as a discard. This resulted in a bit more work for the compiler team but really not that much. The language moved forward with a new feature and 99% of developers don't have to worry about this little quirk.

**Comment by marksmeltzer** [comment]
@CyrusNajmabadi I like it so far. Because I am lazy, I made the lexer/parser capable of doing an optional "walk-back" regression if a file fails to parse using the latest language version: if it finds a language version that successfully parses the file, it automagically inserts/updates the language version instruction on the file (typically that will be the previous version, but it depends on how recently I have worked on a given codebase). 

I can easily grep the codebase to find the files that need to be manually updated to the latest version.

My workflow consists of updating old files as I have time, updating the legacy syntax issues, and then removing the version instruction to bring the file back to latest. So far that workflow has proven itself quite economical. 

It allows me to try funky new ideas for syntax in specific use case in a feature branch without affecting the rest of my codebase. Often what seemed like a good idea reveals itself to be a bad idea once I start retrofitting it other places. In those cases, I'll either deprecate the new feature using the same mechanism or I revert the relevant changes in the feature branch (whichever is cheaper to do).

I think a concept like that is of general utility.

When I go live, I'll probably shift and have new features trickle down from explicit opt-in to a graded set of "insiders" or "futures" release candidates, and then shift the walk-back mechanism to also support downgrading or removing the version instruction when a given file that used to require opting in to future support becomes part of latest, etc.
